proxy/envoy/ingress.go

```diff
@@ -15,256 +15,222 @@
 package envoy
 
 import (
+	"context"
+	"crypto/sha256"
 	"fmt"
 	"io/ioutil"
+	"net/http"
 	"sort"
 	"strconv"
 	"time"
 
 	"github.com/golang/glog"
-	"github.com/golang/protobuf/proto"
 	"github.com/hashicorp/go-multierror"
 
-	"istio.io/api/proxy/v1/config"
+	proxyconfig "istio.io/api/proxy/v1/config"
 	"istio.io/manager/model"
 	"istio.io/manager/proxy"
 )
 
+const (
+	ingressNode = "ingress"
+	certFile    = "/etc/tls.crt"
+	keyFile     = "/etc/tls.key"
+)
+
 type ingressWatcher struct {
 	agent   proxy.Agent
-	ctl     model.Controller
-	context *IngressConfig
+	secrets model.SecretRegistry
+	mesh    *proxyconfig.ProxyMeshConfig
 }
 
 // NewIngressWatcher creates a new ingress watcher instance with an agent
-func NewIngressWatcher(ctl model.Controller, context *IngressConfig) (Watcher, error) {
-	agent := proxy.NewAgent(runEnvoy(context.Mesh, "ingress"), cleanupEnvoy(context.Mesh), 10, 100*time.Millisecond)
-
+func NewIngressWatcher(mesh *proxyconfig.ProxyMeshConfig, secrets model.SecretRegistry) (Watcher, error) {
+	agent := proxy.NewAgent(runEnvoy(mesh, ingressNode), cleanupEnvoy(mesh), 10, 100*time.Millisecond)
 	out := &ingressWatcher{
 		agent:   agent,
-		ctl:     ctl,
-		context: context,
+		secrets: secrets,
+		mesh:    mesh,
 	}
+	return out, nil
+}
 
-	if err := ctl.AppendConfigHandler(model.IngressRule, func(model.Key, proto.Message, model.Event) {
-		out.reload()
-	}); err != nil {
-		return nil, err
-	}
+func (w *ingressWatcher) Run(stop <-chan struct{}) {
+	ctx, cancel := context.WithCancel(context.Background())
+	go w.agent.Run(stop)
+	go func() {
+		<-stop
+		glog.V(2).Info("Ingress watcher terminating...")
+		cancel()
+	}()
+
+	client := &http.Client{Timeout: convertDuration(w.mesh.ConnectTimeout)}
+	url := fmt.Sprintf("http://%s/v1alpha/secret/%s/%s",
+		w.mesh.DiscoveryAddress, w.mesh.IstioServiceCluster, ingressNode)
+	initial := true
+	for {
+		tls, err := fetchSecret(ctx, client, url, w.secrets)
+		if err != nil && !initial {
+			glog.Warning(err)
+		} else {
+			config := generateIngress(w.mesh, tls, certFile, keyFile)
```

Comment:
I think this is trying to only log a warning on the first attempt and only use `tls` if fetchSecret is successfully? But the code here will use `tls` on non-initial attempts when `err != nil`. Maybe something like the following?

```golang
if err != nil {
    if !warned {
        warned = true
    } // else skip warning to avoid flooding logs, maybe add a counter?
} else {
    warned = false // reset warning log whenever fetchSecret is successfully.
    // schedule update
}
```